syntax = "proto3";

package Swoq.Interface;

service Player {
	rpc Register(RegisterRequest) returns (RegisterResponse) {};
}

service Training {
	rpc StartGame(StartRequest) returns (StartResponse) {};
	rpc Move(MoveRequest) returns (MoveResponse) {};
}

enum Result {
	OK = 0;
	
	INTERNAL_ERROR = 1;
	PLAYER_ALREADY_REGISTERED = 2;
	MOVE_NOT_ALLOWED = 3;
}

enum Direction {
	NORTH = 0;
	EAST = 1;
	SOUTH = 2;
	WEST = 3;
}

// Keep synchronized with Swoq.Server.Game.GetCellState()
enum Cell {
	UNKNOWN = 0;
	EMPTY = 1;
	PLAYER = 2;
	WALL = 3;
	EXIT = 4;
}


message State {
	repeated Cell map = 2;
	bool finished = 3;
}

message RegisterRequest {
	string playerName = 1;
}

message RegisterResponse {
	Result result = 1;
	optional string playerId = 2;
}

message StartRequest {
	string playerId = 1;
}

message StartResponse {
	Result result = 1;
	optional string gameId = 2;
	optional int32 height = 3;
	optional int32 width = 4;
	optional State state = 5;
}

message MoveRequest {
	string gameId = 1;
	Direction direction = 2;
}

message MoveResponse {
	Result result = 1;
	optional State state = 2;
}